integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java

```diff
@@ -174,6 +183,44 @@ void requestFiltersCanRespondWithoutProxying(KafkaCluster cluster, Admin admin)
         }
     }
 
+    public static Stream<Arguments> requestFilterCanShortCircuitResponse() {
+        return Stream.of(
+                Arguments.of("synchronous with close", true, ForwardingStyle.SYNCHRONOUS),
+                Arguments.of("synchronous without close", false, ForwardingStyle.SYNCHRONOUS),
+                Arguments.of("asynchronous with close", true, ForwardingStyle.ASYNCHRONOUS_DELAYED),
+                Arguments.of("asynchronous without close", true, ForwardingStyle.ASYNCHRONOUS_REQUEST_TO_BROKER));
+    }
+
+    @ParameterizedTest(name = "{0}")
+    @MethodSource
+    public void requestFilterCanShortCircuitResponse(String name, boolean closeConnection, ForwardingStyle forwardingStyle) {
+        var rejectFilter = new FilterDefinitionBuilder("CreateTopicReject")
+                .withConfig("withCloseConnection", closeConnection,
+                        "forwardingStyle", forwardingStyle)
+                .build();
+        try (var tester = mockKafkaKroxyliciousTester((mockBootstrap) -> proxy(mockBootstrap).addToFilters(rejectFilter));
+                var requestClient = tester.multiRequestClient()) {
+
+            if (forwardingStyle == ForwardingStyle.ASYNCHRONOUS_REQUEST_TO_BROKER) {
+                tester.addMockResponseForApiKey(new Response(LIST_GROUPS, LIST_GROUPS.latestVersion(), new ListGroupsResponseData()));
+            }
+
+            var createTopic = new CreateTopicsRequestData();
+            createTopic.topics().add(new CreateTopicsRequestData.CreatableTopic().setName("foo"));
+
+            var response = requestClient.getSync(new Request(CREATE_TOPICS, CREATE_TOPICS.latestVersion(), "client", createTopic));
+            assertThat(response.message()).isInstanceOf(CreateTopicsResponseData.class);
+
+            var responseMessage = (CreateTopicsResponseData) response.message();
+            assertThat(responseMessage.topics()).allMatch(p -> p.errorCode() == Errors.INVALID_TOPIC_EXCEPTION.code(),
+                    "response contains topics without the expected errorCode");
+
+            await().atMost(Duration.ofSeconds(5))
+                    .untilAsserted(() -> assertThat(requestClient.isOpen()).isEqualTo(!closeConnection));
```

Comment:
```suggestion
                    .untilAsserted(() -> assertThat(requestClient.isOpen()).isNotEqualTo(closeConnection));
```

I'm tempted to say we should invert the `closeConnection` variable so this becomes 
```suggestion
                    .untilAsserted(() -> assertThat(requestClient.isOpen()).isEqualTo(expectConnectionOpen));
```

Reply:
I like the fact that the test is in terms of the public API.  I mean the FilterResultBuilder talks in terms of the `withCloseConnection`, so the test should too. That made me prefer to keep the test's interface the same.  

I extracted a variable `expectConnectionOpen` to reduce the mental gymnastics.